### PR TITLE
optimize KeyGroupIterator

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/KeyGroupIterator.java [hunk 1]
```diff
@@ -17,69 +17,92 @@
...
+        iterators.stream().map(Iterators::peekingIterator).collect(Collectors.toList());
     this.keyFn = keyFn;
     this.keyComparator = keyComparator;
   }
 
   @Override
   public boolean hasNext() {
     checkState();
-    return iterator.hasNext();
+    return iterators.stream().anyMatch(it -> it.hasNext() && it.peek() != null);
```
Comment (nevillelyh) [inline scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/KeyGroupIterator.java]: I'm not sure this 100% correct. The `!= null` condition is basically treating `null` key as not a real key but it also seems that our E2E test doesn't cover cases with multiple `null` keys that might mess up the iterator state?

Comment (nevillelyh) [reply scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/KeyGroupIterator.java]: This might be a red herring and caused by the bug in `JsonFileOperations`.
